docs/topics/virtual-warping.rst

```diff
@@ -0,0 +1,53 @@
+Virtual Warping
+===============
+
+Rasterio has a class that abstracts many of the details of raster warping by
+using an in-memory `Warped VRT
+<http://www.gdal.org/gdal_vrttut.html#gdal_vrttut_warped>`__.
+
+To virtually warp the Rasterio test dataset to EPSG:3857 and extract pixels
+corresponding to its central zoom 9 tile, do the following.
+
+.. code-block:: python
+
+  import mercantile
+
+  from rasterio.io import VirtualWarpedFile
+  from rasterio.enums import Resampling
+
+
+  with VirtualWarpedFile('tests/data/RGB.byte.tif', dst_crs='EPSG:3857',
+                         resampling=Resampling.bilinear).open() as src:
+
+      # Determine the destination tile and its mercator bounds using
+      # functions from the mercantile module.
+      dst_tile = mercantile.tile(*src.lnglat(), 9)
+      left, top = mercantile.xy(*mercantile.ul(*dst_tile))
+      right, bottom = mercantile.xy(*mercantile.ul(
+          mercantile.Tile(dst_tile.x + 1, dst_tile.y + 1, dst_tile.z)))
+
+      # Determine the window to use in reading from the dataset.
+      dst_window = src.window(left, bottom, right, top)
+
+      # Read into a 3 x 512 x 512 array. Our output tile will be
+      # 512 wide x 512 tall.
+      data = src.read(window=dst_window, out_shape=(3, 512, 512))
```

Comment:
This assumes that the source dataset overlaps the desired window. If not, the output shape will still be `(3, 512, 512)`, but the data within that array will be stretched if the target window extends beyond one of the dimensions of the source.